### PR TITLE
Fix bootstrap

### PR DESCRIPTION
**Pull request by odersky**
Fixes needed so that dotc can selectively compile some parts of itself, using TASTY as the pickled symbol format. Based on #1243.

### CODE REVIEW

File: src/dotty/tools/dotc/core/Symbols.scala [hunk 1]
```diff
@@ -398,10 +398,10 @@ object Symbols {
 
     /** Subclass tests and casts */
     final def isTerm(implicit ctx: Context): Boolean =
-      (if(isDefinedInCurrentRun) lastDenot else denot).isTerm
+      (if (defRunId == ctx.runId) lastDenot else denot).isTerm
```
Comment (smarter) [inline src/dotty/tools/dotc/core/Symbols.scala]: Why not change the definition of `isDefinedInCurrentRun` ?

File: src/dotty/tools/dotc/core/Contexts.scala [hunk 2]
```diff
@@ -248,7 +248,7 @@ object Contexts {
       withPhase(phase.id)
 
     final def withPhaseNoLater(phase: Phase) =
-      if (ctx.phase.id > phase.id) withPhase(phase) else ctx
+      if (phase.exists && ctx.phase.id > phase.id) withPhase(phase) else ctx
```
Comment (smarter) [inline src/dotty/tools/dotc/core/Contexts.scala]: Does this happen in practice?

File: src/dotty/tools/dotc/core/Symbols.scala (hunk 1 above)
Comment (odersky) [inline src/dotty/tools/dotc/core/Symbols.scala]: Because it's used elsewhere to mean: In a compilation unit that's currently being compiled.

File: src/dotty/tools/dotc/core/Contexts.scala (hunk 2 above)
Comment (odersky) [inline src/dotty/tools/dotc/core/Contexts.scala]: Yes, there are tests that break (the tests don't start a full compiler).

**Comment by odersky** [comment]
We now can selectively compile individual dotc files and directories using Tasty to access the information of the other parts of dotc. #1274 demonstrates what I did to verify this.

**Comment by odersky** [comment]
Rebased to master

File: src/dotty/tools/dotc/transform/TreeTransform.scala [hunk 3]
```diff
@@ -174,32 +174,22 @@ object TreeTransforms {
  }
 
   /** A helper trait to transform annotations on MemberDefs */
-  trait AnnotationTransformer extends MiniPhaseTransform with InfoTransformer {
+  trait AnnotationTransformer extends MiniPhaseTransform with DenotTransformer {
```
Comment (DarkDimius) [inline src/dotty/tools/dotc/transform/TreeTransform.scala]: Why is this change needed?
As annotation transformers also change types, isn't every annotation transformer and info-transformer?

File: src/dotty/tools/dotc/core/tasty/TastyFormat.scala [hunk 4]
```diff
@@ -485,4 +485,15 @@ object TastyFormat {
     case PRIVATEqualified => "PRIVATEqualified"
     case PROTECTEDqualified => "PROTECTEDqualified"
   }
+
+  /** @return If non-negative, the number of leading references of a length/trees entry.
+   *          If negative, minus the number of leading non-reference trees.
+   */
+  def numRefs(tag: Int) = tag match {
```
Comment (DarkDimius) [inline src/dotty/tools/dotc/core/tasty/TastyFormat.scala]: This method seem to only be used by tasty-unpickler. Maybe move it there?

File: src/dotty/tools/dotc/core/tasty/TreePickler.scala [hunk 5]
```diff
@@ -217,8 +217,8 @@ class TreePickler(pickler: TastyPickler) {
     case tpe: RefinedType =>
       writeByte(REFINEDtype)
       withLength {
-        pickleType(tpe.parent)
```
Comment (DarkDimius) [inline src/dotty/tools/dotc/core/tasty/TreePickler.scala]: This is an incompatible change. Why was the order changed?

File: src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala [hunk 6]
```diff
@@ -145,19 +209,25 @@ class TreeUnpickler(reader: TastyReader, tastyName: TastyName.Table) {
...
-          // Create a stub; owner might be wrong but will be overwritten later.
-          forkAt(addr).createSymbol()
-          val sym = symAtAddr(addr)
-          ctx.log(i"forward reference to $sym")
-          stubs += sym
-          sym
-      }
+    def readSymRef()(implicit ctx: Context): Symbol = symbolAt(readAddr())
+
+    /** The symbol at given address; createa new one if none exists yet */
```
Comment (DarkDimius) [inline src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala]: `createa` : typo

File: src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala [hunk 7]
```diff
@@ -393,44 +476,31 @@ class TreeUnpickler(reader: TastyReader, tastyName: TastyName.Table) {
...
-                  preExisting
-                case none =>
-                  ctx.newNakedSymbol(start.index)
-              }
-              val denot = ctx.SymDenotation(symbol = sym, owner = ctx.owner, name, flags, completer, privateWithin)
-              sym.denot = denot
-              sym
-            }
-        } // TODO set position
+              ctx.newClassSymbol(ctx.owner, name.asTypeName, flags, completer, privateWithin, coord = start.index)
```
Comment (DarkDimius) [inline src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala]: The only difference between branches is the name. Maybe try:
```scala
val symName = if (isClass) name.asTypeName else name
ctx.newClassSymbol(ctx.owner, name.asTypeName, flags, completer, privateWithin, coord = start.index)

```

File: src/dotty/tools/dotc/core/tasty/TreePickler.scala (hunk 5 above)
Comment (odersky) [inline src/dotty/tools/dotc/core/tasty/TreePickler.scala]: To harmonize the layout. Non-trees now always come before trees in Tasy entries (except for some cases where trailing non-trees are optional as explained in numRefs.

File: src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala (hunk 7 above)
Comment (odersky) [inline src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala]: And whether create a symbol or a class symbol no? That prevented me from merging the branches.

File: src/dotty/tools/dotc/core/tasty/TastyFormat.scala (hunk 4 above)
Comment (odersky) [inline src/dotty/tools/dotc/core/tasty/TastyFormat.scala]: I think it is essentially an information about the format. So the place looks right to me. Otherwise put, if we change the format of some entries, this method needs to be updated in sync. With the information provided by it, we enable generic traversals of Tasty trees in other (future) modules as well.

File: src/dotty/tools/dotc/transform/TreeTransform.scala (hunk 3 above)
Comment (odersky) [inline src/dotty/tools/dotc/transform/TreeTransform.scala]: Currently that's not true, but I see you argument. I..e if we left out the self type transform from FirstTransform it would not be an InfoTransformer. And that transform is in FirstTransform only by accident, it could well be elsewhere.

*However* one could argue that every annotation transform should affect annotated types, and annotated types can be infos of symbols so in that sense every annotation transform *is* an info transform.

*But* that would require a complete typemap of every type in every info transform, and we know this leads to cyclic references. So we can't do this at present.

Bottom line: When in doubt decouple the functionalities.

File: src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala (hunk 7 above)
Comment (DarkDimius) [inline src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala]: My bad. You're right.

**Comment by odersky** [comment]
ready to merge?

**Comment by DarkDimius** [comment]
LGTM
